benchmark-framework/src/main/java/io/openmessaging/benchmark/WorkloadGenerator.java

```diff
@@ -355,38 +357,44 @@ private TestResult generateLoad(List<BenchmarkProducer> producers, long testDura
         long startTime = System.nanoTime();
         this.testCompleted = false;
 
-        executor.submit(() -> {
-            try {
-                byte[] payloadData = new byte[workload.messageSize];
-
-                // Send messages on all topics/producers
-                while (!testCompleted) {
-                    for (int i = 0; i < producers.size(); i++) {
-                        BenchmarkProducer producer = producers.get(i);
-                        rateLimiter.acquire();
-
-                        final long sendTime = System.nanoTime();
-
-                        producer.sendAsync(payloadData).thenRun(() -> {
-                            messagesSent.increment();
-                            totalMessagesSent.increment();
-                            bytesSent.add(payloadData.length);
-
-                            long latencyMicros = TimeUnit.NANOSECONDS.toMicros(System.nanoTime() - sendTime);
-                            recorder.recordValue(latencyMicros);
-                            cumulativeRecorder.recordValue(latencyMicros);
-
-                        }).exceptionally(ex -> {
-                            log.warn("Write error on message", ex);
-                            System.exit(-1);
-                            return null;
-                        });
+        int processors = Runtime.getRuntime().availableProcessors();
+        Collections.shuffle(producers);
+
+        // Divide the producers across multiple different threads
+        for (List<BenchmarkProducer> producersPerThread : Lists.partition(producers, processors)) {
+            executor.submit(() -> {
+                try {
+                    byte[] payloadData = new byte[workload.messageSize];
```

Comment:
You can probably have this outside the executor since this payload is always the same